saw-core-what4/src/Verifier/SAW/Simulator/What4/FirstOrder.hs

```diff
@@ -111,13 +113,30 @@ groundToFOV BaseRealRepr    _         = Left "Real is not FOV"
 groundToFOV BaseComplexRepr         _ = Left "Complex is not FOV"
 groundToFOV (BaseStringRepr _)      _ = Left "String is not FOV"
 groundToFOV (BaseFloatRepr _)       _ = Left "Floating point is not FOV"
-groundToFOV (BaseArrayRepr (Empty :> ty) b) _
-  | Right fot1 <- typeReprToFOT ty
-  , Right fot2 <- typeReprToFOT b
-  = pure $ FOVArray fot1 fot2
-groundToFOV (BaseArrayRepr _idx _b) _ = Left "Unsupported FOV Array"
+groundToFOV (BaseArrayRepr _ _) (ArrayMapping _) =
+    Left "Unsupported FOV Array (values only available by lookup)"
+groundToFOV (BaseArrayRepr (Empty :> ty_idx) ty_val) (ArrayConcrete dfl values) = do
+    ty_idx' <- typeReprToFOT ty_idx
+    dfl' <- groundToFOV ty_val dfl
+    let convert mvalues' idx v = do
+          values' <- mvalues'
+          idx' <- indexToFOV ty_idx idx
+          v' <- groundToFOV ty_val v
+          return $ Map.insert idx' v' values'
+    values' <- Map.foldlWithKey convert (pure Map.empty) values
```

Comment:
If you use [`ifoldlM`](https://hackage.haskell.org/package/indexed-traversable-0.1.4/docs/Data-Foldable-WithIndex.html#v:ifoldlM) (from the `indexed-traversable` package, which `saw-script` already depends on), then you can avoid needing the extra `pure` here.